app/6502.c

```diff
@@ -1,175 +1,45 @@
 #include <stdio.h>
-#include <stdlib.h>
-#include <SDL2/SDL.h>
-#include "SDL_pixels.h"
-
-struct State
-{
-  int key_quit;
-  int key_up;
-  int key_left;
-  int key_right;
-  int key_down;
-};
-
-struct Point
-{
-  int x;
-  int y;
-};
-
-SDL_Window* init_window()
-{
-  // returns zero on success else non-zero
-  if (SDL_Init(SDL_INIT_EVERYTHING) != 0) {
-    printf("error initializing SDL: %s\n", SDL_GetError());
-  }
-
-  return SDL_CreateWindow("6502",
-      SDL_WINDOWPOS_CENTERED,
-      SDL_WINDOWPOS_CENTERED,
-      1000, 1000, 0); 
+#include "cpu.h"
+#include "sdl.h"
+
+void initSimulator (struct CPU * cpu, struct BUS * bus, int initial_address, char * programPath) {
+  writeBus(0xFFFC, initial_address & 0xff, bus);
+  writeBus(0xFFFD, (initial_address >> 8), bus);
+  resetCpu(cpu, bus);
+  writeFileBus(programPath, initial_address, bus);
 }
 
-struct State event_manager(struct State state) {
-  struct State local_state = state;
-  SDL_Event event;
-
-  while (SDL_PollEvent(&event)) {
-
-    if (event.type == SDL_QUIT)
-      local_state.key_quit = 1;
+void loop(SDL_Renderer *rend, struct CPU * cpu, struct BUS * bus) {
+  int current_key = SDL_FIRSTEVENT;
 
-    if (event.type == SDL_KEYDOWN)
-    {
-      switch (event.key.keysym.scancode) {
-        case SDL_SCANCODE_W:
-        case SDL_SCANCODE_UP:
-          local_state.key_up = 1;
-          break;
-        case SDL_SCANCODE_A:
-        case SDL_SCANCODE_LEFT:
-          local_state.key_left = 1;
-          break;
-        case SDL_SCANCODE_S:
-        case SDL_SCANCODE_DOWN:
-          local_state.key_down = 1;
-          break;
-        case SDL_SCANCODE_D:
-        case SDL_SCANCODE_RIGHT:
-          local_state.key_right = 1;
-          break;
-        default:
-          break;
-      }
-    }
-    if (event.type == SDL_KEYUP)
-    {
-      switch (event.key.keysym.scancode) {
-        case SDL_SCANCODE_W:
-        case SDL_SCANCODE_UP:
-          local_state.key_up = 0;
-          break;
-        case SDL_SCANCODE_A:
-        case SDL_SCANCODE_LEFT:
-          local_state.key_left = 0;
-          break;
-        case SDL_SCANCODE_S:
-        case SDL_SCANCODE_DOWN:
-          local_state.key_down = 0;
-          break;
-        case SDL_SCANCODE_D:
-        case SDL_SCANCODE_RIGHT:
-          local_state.key_right = 0;
-          break;
-        default:
-          break;
-      }
-    }
+  while (current_key != SDL_QUIT) {
+    current_key = readInputSdl();
+    bus->input = (unsigned char)current_key;
+    drawScreenSdl(rend, bus->display);
```

Comment:
Render first is some kind of rule?
I mean, the clock came after and we always draw (or redraw) before...

Reply:
No, it was arbitrary I didn't though about, what you think after or before ? Maybe after the clock right?

Reply:
Well, if SDL persists what it drawed before, probably after clock is a good place to be

Reply:
I do a clean screen every frame 😬